subsys/bluetooth/host/gatt.c

```diff
@@ -4650,7 +4650,23 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
 
 void bt_gatt_cancel(struct bt_conn *conn, void *params)
 {
-	bt_att_req_cancel(conn, params);
+	struct bt_att_req *req;
+	bt_att_func_t func;
+
+	k_sched_lock();
+
+	req = bt_att_find_req_by_user_data(conn, params);
+
+	if (req) {
+		func = req->func;
+		bt_att_req_cancel(conn, req);
+	}
+
+	k_sched_unlock();
+
+	if (req) {
+		func(conn, BT_ATT_ERR_UNLIKELY, NULL, 0, params);
+	}
```

Comment:
Minor comment: It feels unnecessary to store `func` in a variable instead of doing
```suggestion
	struct bt_att_req *req;

	k_sched_lock();

	req = bt_att_find_req_by_user_data(conn, params);
	if (req) {
		bt_att_req_cancel(conn, req);
	}

	k_sched_unlock();

	if (req) {
		req->func(conn, BT_ATT_ERR_UNLIKELY, NULL, 0, params);
	}
```

Reply:
`func` must be saved. `bt_att_req_cancel` calls `bt_att_req_free`. (A bit surprising, I agree. It's not what the function name says.)

It's a common pattern in this module that we free the `req` before calling into callbacks. This because the callback might result in the application performing a new request, which will allocate a `req` again.

Reply:
By `free` I suppose you mean reset, right? 

Otherwise we shouldn't do a check for `if (req)` { func(...);}`. 

If we store the func, then perhaps it should be changed to 
```suggestion
	struct bt_att_req *req;
	bt_att_func_t func = NULL;
	
	k_sched_lock();

	req = bt_att_find_req_by_user_data(conn, params);
	if (req) {
		func = req->func;
		bt_att_req_cancel(conn, req);
	}

	k_sched_unlock();

	if (func) {
		func(conn, BT_ATT_ERR_UNLIKELY, NULL, 0, params);
	}
```

Reply:
I made the two ifs have the same predicate for readability. My intention is the visual appearance that it is the same conditional, interleaved by a unconditional call to `k_sched_unlock`.